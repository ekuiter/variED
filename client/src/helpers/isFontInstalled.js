import memoize from './memoize';

/**
 * JavaScript code to detect available availability of a
 * particular font in a browser using JavaScript and CSS.
 * Author : Lalit Patel
 * Website: http://www.lalit.org/lab/javascript-css-font-detect/
 */

const baseFonts = ['monospace', 'sans-serif', 'serif'],
    testString = 'mmmmmmmmmmlli',
    testSize = '72px',
    defaultWidth = {},
    defaultHeight = {},
    body = document.getElementsByTagName("body")[0],
    span = document.createElement("span");

span.style.fontSize = testSize;
span.innerHTML = testString;
baseFonts.forEach(baseFont => {
    span.style.fontFamily = baseFont;
    body.appendChild(span);
    defaultWidth[baseFont] = span.offsetWidth;
    defaultHeight[baseFont] = span.offsetHeight;
    body.removeChild(span);
});

export default memoize(function(font) {
    let detected = false;
    baseFonts.forEach(baseFont => {
        span.style.fontFamily = `${font},${baseFont}`;
        body.appendChild(span);
        detected = detected || ((span.offsetWidth !== defaultWidth[baseFont] || span.offsetHeight !== defaultHeight[baseFont]));
        body.removeChild(span);
    });
    return detected;
});