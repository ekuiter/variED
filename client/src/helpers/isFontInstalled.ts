/**
 * JavaScript code to detect available availability of a
 * particular font in a browser using JavaScript and CSS.
 * Author : Lalit Patel
 * Website: http://www.lalit.org/lab/javascript-css-font-detect/
 */

import memoize from './memoize';

const baseFonts = ['monospace', 'sans-serif', 'serif'],
    testString = 'mmmmmmmmmmlli',
    testSize = '72px',
    defaultWidth = {},
    defaultHeight = {},
    body = document.getElementsByTagName('body')[0],
    span = document.createElement('span');

span.style.fontSize = testSize;
span.innerHTML = testString;
baseFonts.forEach(baseFont => {
    span.style.fontFamily = baseFont;
    body.appendChild(span);
    defaultWidth[baseFont] = span.offsetWidth;
    defaultHeight[baseFont] = span.offsetHeight;
    body.removeChild(span);
});

/**
 * Returns whether a font is installed on the system.
 * This is an expensive operation, so we memoize its results for further use.
 * We do not expect the system's installed functions to change while the application is running.
 * @param font font family to detect, should NOT be a CSS-like list of font families
 */
export default <(font: string) => boolean>memoize(function(font) {
    let detected = false;
    baseFonts.forEach(baseFont => {
        span.style.fontFamily = `${font},${baseFont}`;
        body.appendChild(span);
        detected = detected || ((span.offsetWidth !== defaultWidth[baseFont] || span.offsetHeight !== defaultHeight[baseFont]));
        body.removeChild(span);
    });
    return detected;
});