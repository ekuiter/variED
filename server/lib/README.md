## Libraries

This folder contains the FeatureIDE core library
(`plugins/de.ovgu.featureide.fm.core`) and its dependencies.

The core library (`de.ovgu.featureide.lib.fm.jar`) was built from [this
commit](https://github.com/FeatureIDE/FeatureIDE/tree/aec4288772d3a1cfb89a0cd14a9abb6281daedf5/plugins/de.ovgu.featureide.fm.core/library).

The additional libraries can be found
[here](https://github.com/FeatureIDE/FeatureIDE/tree/aec4288772d3a1cfb89a0cd14a9abb6281daedf5/plugins/de.ovgu.featureide.fm.core/lib).
