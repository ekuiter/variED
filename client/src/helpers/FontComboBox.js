import isFontInstalled from './isFontInstalled';
import {ComboBox} from 'office-ui-fabric-react/lib/ComboBox';
import React from 'react';
import i18n from '../i18n';
import constants from '../constants';
import PropTypes from 'prop-types';

export default class extends React.Component {
    static propTypes = {
        selectedFont: PropTypes.string.isRequired,
        fonts: PropTypes.arrayOf(PropTypes.string),
        onChange: PropTypes.func.isRequired
    };

    static defaultProps = {fonts: constants.helpers.fontComboBox.suggestedFonts};
    state = {errorMessage: null};

    onChange = (option, index, value) => {
        const font = value || (option && option.key);
        if (!font)
            return;
        if (isFontInstalled(font))
            this.props.onChange(font);
        else
            this.setState({errorMessage: i18n.t('panels.settingsPanel.errors.fontNotInstalled')});
    };

    render() {
        let {selectedFont, fonts, ...props} = this.props;
        fonts = fonts
            .filter(isFontInstalled)
            .map(font => ({key: font, text: font, fontFamily: `'${font}'`}));

        return (
            <ComboBox
                text={this.state.errorMessage ? null : selectedFont}
                allowFreeform={true}
                autoComplete="on"
                options={fonts}
                onRenderOption={({text, fontFamily}) =>
                    <span className="ms-ComboBox-optionText" style={{fontFamily}}>{text}</span>}
                errorMessage={this.state.errorMessage}
                {...props}
                onChange={this.onChange}/>
        );
    }
}