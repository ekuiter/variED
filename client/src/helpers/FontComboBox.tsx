/**
 * A Fabric combo box that allows to choose a font installed on the user's system.
 */

import isFontInstalled from './isFontInstalled';
import {ComboBox, IComboBox, IComboBoxOption, IComboBoxProps} from 'office-ui-fabric-react/lib/ComboBox';
import React from 'react';
import i18n from '../i18n';
import constants from '../constants';

interface Props {
    selectedFont: string,
    fonts?: string[],
    onChange: (font: string) => void,
    comboBoxProps?: Partial<IComboBoxProps>
};

interface State {
    errorMessage?: string
};

export default class extends React.Component<Props, State> {
    static defaultProps: Partial<Props> = {fonts: constants.helpers.fontComboBox.suggestedFonts};
    state: State = {errorMessage: undefined};

    onChange = (_event: React.FormEvent<IComboBox>, option?: IComboBoxOption, index?: number, value?: string) => {
        const font = value || (option && option.key);
        if (typeof font !== 'string')
            return;
        if (isFontInstalled(font)) {
            this.setState({errorMessage: undefined});
            this.props.onChange(font);
        } else
            this.setState({errorMessage: i18n.t('overlays.settingsPanel.errors.fontNotInstalled')});
    };

    render() {
        let {selectedFont, fonts, comboBoxProps} = this.props;
        const options = fonts!
            .filter(isFontInstalled)
            .map(font => ({key: font, text: font}));

        return (
            <ComboBox
                text={this.state.errorMessage ? undefined : selectedFont}
                allowFreeform={true}
                autoComplete="on"
                options={options}
                onRenderOption={(props?: {text: string}) =>
                    <span className="ms-ComboBox-optionText" style={{fontFamily: `'${props!.text}'`}}>{props!.text}</span>}
                errorMessage={this.state.errorMessage}
                {...comboBoxProps}
                onChange={this.onChange}/>
        );
    }
}