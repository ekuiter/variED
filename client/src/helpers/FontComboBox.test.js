import React from 'react';
import FontComboBox from './FontComboBox';
import {shallow} from 'enzyme';
import {ComboBox} from 'office-ui-fabric-react/lib/ComboBox';
import isFontInstalled from './isFontInstalled';

jest.mock('./isFontInstalled');

describe('FontComboBox', () => {
    const fontComboBox = (onChange = jest.fn()) => shallow(
        <FontComboBox
            selectedFont="Arial"
            onChange={onChange}/>
    );

    it('renders correctly', () => {
        isFontInstalled.mockReturnValue(true);
        const wrapper = fontComboBox();
        expect(wrapper.find(ComboBox).prop('text')).toBe('Arial');
        expect(wrapper.find(ComboBox).prop('options'))
            .toContainEqual({key: 'Arial', text: 'Arial', fontFamily: '\'Arial\''});
        expect(wrapper).toMatchSnapshot();
    });

    it('renders font options', () => {
        isFontInstalled.mockReturnValue(true);
        const wrapper = fontComboBox();
        expect(wrapper.find(ComboBox).prop('onRenderOption')(
            {key: 'Arial', text: 'Arial', fontFamily: '\'Arial\''})).toMatchSnapshot();
    });

    it('changes a font if it is installed', () => {
        isFontInstalled.mockReturnValue(true);
        const onChange = jest.fn(),
            wrapper = fontComboBox(onChange);
        wrapper.find(ComboBox).simulate('change', {key: 'Verdana'});
        expect(onChange).toBeCalledWith('Verdana');
    });

    it('displays an error if a font is not installed', () => {
        isFontInstalled.mockReturnValue(false);
        const onChange = jest.fn(),
            wrapper = fontComboBox(onChange);
        expect(wrapper.find(ComboBox).prop('errorMessage')).toBeFalsy();
        wrapper.find(ComboBox).simulate('change', {key: 'Verdana'});
        expect(onChange).not.toBeCalledWith('Verdana');
        expect(wrapper.find(ComboBox).prop('errorMessage')).toBeTruthy();
    });

    it('does nothing if no font is provided', () => {
        isFontInstalled.mockReturnValue(true);
        const onChange = jest.fn(),
            wrapper = fontComboBox(onChange);
        wrapper.find(ComboBox).simulate('change');
        expect(onChange).not.toBeCalled();
    });
});