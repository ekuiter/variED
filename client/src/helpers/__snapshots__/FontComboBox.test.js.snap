// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FontComboBox renders correctly 1`] = `
<CustomizedComboBox
  allowFreeform={true}
  autoComplete="on"
  buttonIconProps={
    Object {
      "iconName": "ChevronDown",
    }
  }
  errorMessage={null}
  onChange={[Function]}
  onRenderOption={[Function]}
  options={
    Array [
      Object {
        "fontFamily": "'Arial'",
        "key": "Arial",
        "text": "Arial",
      },
      Object {
        "fontFamily": "'Calibri'",
        "key": "Calibri",
        "text": "Calibri",
      },
      Object {
        "fontFamily": "'Candara'",
        "key": "Candara",
        "text": "Candara",
      },
      Object {
        "fontFamily": "'Century Gothic'",
        "key": "Century Gothic",
        "text": "Century Gothic",
      },
      Object {
        "fontFamily": "'Courier New'",
        "key": "Courier New",
        "text": "Courier New",
      },
      Object {
        "fontFamily": "'Futura'",
        "key": "Futura",
        "text": "Futura",
      },
      Object {
        "fontFamily": "'Garamond'",
        "key": "Garamond",
        "text": "Garamond",
      },
      Object {
        "fontFamily": "'Geneva'",
        "key": "Geneva",
        "text": "Geneva",
      },
      Object {
        "fontFamily": "'Georgia'",
        "key": "Georgia",
        "text": "Georgia",
      },
      Object {
        "fontFamily": "'Helvetica'",
        "key": "Helvetica",
        "text": "Helvetica",
      },
      Object {
        "fontFamily": "'Helvetica Neue'",
        "key": "Helvetica Neue",
        "text": "Helvetica Neue",
      },
      Object {
        "fontFamily": "'Lucida Grande'",
        "key": "Lucida Grande",
        "text": "Lucida Grande",
      },
      Object {
        "fontFamily": "'Palatino'",
        "key": "Palatino",
        "text": "Palatino",
      },
      Object {
        "fontFamily": "'Segoe UI'",
        "key": "Segoe UI",
        "text": "Segoe UI",
      },
      Object {
        "fontFamily": "'Tahoma'",
        "key": "Tahoma",
        "text": "Tahoma",
      },
      Object {
        "fontFamily": "'Times New Roman'",
        "key": "Times New Roman",
        "text": "Times New Roman",
      },
      Object {
        "fontFamily": "'Trebuchet MS'",
        "key": "Trebuchet MS",
        "text": "Trebuchet MS",
      },
      Object {
        "fontFamily": "'Verdana'",
        "key": "Verdana",
        "text": "Verdana",
      },
    ]
  }
  text="Arial"
/>
`;

exports[`FontComboBox renders font options 1`] = `
<span
  className="ms-ComboBox-optionText"
  style={
    Object {
      "fontFamily": "'Arial'",
    }
  }
>
  Arial
</span>
`;
